Created UTC timestamping  functionality

kgrewal26 opened the pull request:
A few things I should probably note:

Not entirely too sure if this would work, I mainly based it off my conversation with Dhruv and Tony.

The timer setup is the thing I'm most concerned about - I'm really not sure if I did it right.

Also, the period of the timer hasn't been set yet

Pop0097 commented on Autopilot/Inc/timeStampingPOGI.hpp:
```diff
@@ -0,0 +1,25 @@
+
+};
+
+#endif
```
If you see this "do not enter" type sign, it means you are missing a neweline character at the end of your file. To fix if you just have to go to the bottom of the file and hit enter once or twice :))

Pop0097 commented on Autopilot/Inc/timeStampingPOGI.hpp:
```diff
@@ -0,0 +1,25 @@
+#include <cstdlib>
+#include "gps.hpp"
+
+class timeStampingPOGI {
```
`TimeStampingPOGI`

Pop0097 replied:
we use upper CammelCase for classes :))

Pop0097 commented on Autopilot/Inc/timeStampingPOGI.hpp:
```diff
@@ -0,0 +1,25 @@
+
+        static timeStampingPOGI* GetInstance();
+        void setGPSTime(GpsData_t* gpsData); //need to change data type depending on what sensorFusion sends me
+        uint32_t getUTCTimeMS();
```
If you could put comments over some of the methods using the following strucuture, that would be great:

```cpp
/**
* Explain what function does
*/
void functionName();
```

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,56 @@
+//just typical singleton stuff
+
+//Allocates and initializes the static data member pointer (not the object itself)
+timeStampingPOGI* timeStampingPOGI::s_Instance = NULL;
```
`NULL` is C, we use `nullptr` in C++ :))

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,56 @@
+//creates the instance if it doesnt exist & returns the pointer to the instance
+timeStampingPOGI* timeStampingPOGI::GetInstance(){
+
+    if(s_Instance == NULL){
```
`nullptr`

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,56 @@
+//call this to receive the milliseconds since midnight (in UTC)
+uint32_t timeStampingPOGI::getUTCTimeMS(){
+    
+    return (recentUTCTime*1000) + __HAL_TIM_GET_COUNTER(&htim4);
```
space around `*` please

Pop0097 commented on Autopilot/boardfiles/Src/tim.c:
```diff
@@ -95,6 +96,42 @@ void MX_TIM11_Init(void)
+  htim4.Instance = TIM4;
+  htim4.Init.Prescaler = 0;
+  htim4.Init.CounterMode = TIM_COUNTERMODE_UP;
+  htim4.Init.Period = 65535; //STILL NEEDS TO BE SET
```
a 4 bit timer cannot go up to 65535. You need at least a 16 bit timer for this (TIM16)

Pop0097 commented on Autopilot/boardfiles/Src/tim.c:
```diff
@@ -95,6 +96,42 @@ void MX_TIM11_Init(void)
+  TIM_IC_InitTypeDef sConfigIC = {0};
+
+  htim4.Instance = TIM4;
+  htim4.Init.Prescaler = 0;
```
I believe prescalar needs to at least be 1.

Also, what period are you trying to achieve?

kgrewal26 replied:
That was one thing I wasnt too sure about yet - I'm gonna think it over tomorrow and recommit

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,56 @@
+}
+
+
+void timeStampingPOGI::setGPSTime(GpsData_t* gpsData){
+
+    //check if its new and set the recentUTCTime to it
+    if (gpsData->timeIsNew){
+        __HAL_TIM_SET_COUNTER(&htim4, 0);
+        recentUTCTime = gpsData->utcTime;
+    }
+
```
The GPS data comes in as `HHMMSS.00` (24 hour clock). You will need to convert it to UTC time

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,56 @@
+//call this to receive the milliseconds since midnight (in UTC)
+uint32_t timeStampingPOGI::getUTCTimeMS(){
+    
+    return (recentUTCTime*1000) + __HAL_TIM_GET_COUNTER(&htim4);
```
Is the counter counting in milliseconds. If so this is a really smart move

kgrewal26 replied:
Yup - that was my intention. I can just return milliseconds since midnight or manipulate it into whatever format is needed.

The only thing I haven't set yet is the period for the timer - gonna likely do that tomorrow.

Pop0097 reviewed (commented):
Good start!

kgrewal26 commented:
Fixed some of the formatting issues. A few things still need to be done:

Changing up the period & switching from tim4 to something else

Will hopefully have that up by tomorrow!

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -1,10 +1,10 @@
 
-#include "timeStampingPOGI.hpp"
+#include "TimeStampingPOGI.hpp"
```
should be `timeStampingPOGI.hpp`. I am surprised this built on your local... what OS are you using?

kgrewal26 replied:
hmm, that's weird. I renamed the file to TimeStampingPOGI on my local end, I guess those changes didn't sync through git?

Pop0097 replied:
Still need to change the file name in this `#include`

kgrewal26 commented:
Switched to timer12 instead of 4 (this way we can count to at least 1000 with some cushion)

I set the period to 0.001 (1000hz0. The idea behind this is each tick represents 1 millisecond. Let me know if this works because I'm not even entirely sure.

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,57 @@
+//just typical singleton stuff
+
+//Allocates and initializes the static data member pointer (not the object itself)
+TimeStampingPOGI* TimeStampingPOGI::s_Instance = nullptr;
```
declare global variables at the top of the file please

Pop0097 replied:
below the #include and #defines tho :))

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,57 @@
+void TimeStampingPOGI::setGPSTime(GpsData_t* gpsData){
+
+    //check if its new and set the recentUTCTime to it
+    if (gpsData->timeIsNew){
+        __HAL_TIM_SET_COUNTER(&htim12, 0);
+        recentUTCTime = gpsData->utcTime;
+    }
```
You still need to make a method to convert the GPS time to UTC time in seconds since midnight

I want that method to do something like this calculator: https://www.tools4noobs.com/online_tools/hh_mm_ss_to_seconds/

kgrewal26 replied:
oops yeah, not sure how but it completely slipped my mind

Pop0097 commented on Autopilot/boardfiles/Src/tim.c:
```diff
@@ -95,6 +96,44 @@ void MX_TIM11_Init(void)
+  htim12.Init.CounterMode = TIM_COUNTERMODE_UP;
+
+  //this period represents 1000hz - i.e every tick represents a millisecond
+  htim12.Init.Period = 0.001; 
```
if you think back to the bootcamp, the period is the value that your timer will count until.

You want your counter to run at 1000 Hz, so that means your `htim12.Init.Period` should be 1000 if you want each tick to be a milisecond. 

Also, you need to set the prescalar correctly. Our CPU (STM32F765ZG) runs at 216 MHz, so your prescalar must be 216000

kgrewal26 replied:
Also not sure how this slipped my mind - I must've autopiloted. (no pun intended)

kgrewal26 replied:
> Thanks for getting this in. I have a few comments.
> 
> Do you have a dev board that you can test this on?

Unfortunately I do not. Is there any way some else can test this for me?

Pop0097 commented on Autopilot/boardfiles/Src/tim.c:
```diff
@@ -95,6 +96,44 @@ void MX_TIM11_Init(void)
 }
 
+/* TIM12 Init Function */
+void MX_TIM12_Init(void)
```
Ignore this if you have already configured this timer using CubeMX

Right now you have created these timers, but I don't think our chip will be able to run them since we haven't configured them. What you will have to do is use CubeMX (https://www.st.com/en/development-tools/stm32cubemx.html) and add your configurations to our `Autopilot_ZeropilotRev1.ioc` file found in `./boardfiles`. 

After configuration, thse methods will be regenerated and you will need to verify that they are correct. You will also need to include the `.ioc` file in this PR. 

![Screen Shot 2021-06-18 at 6 21 44 PM](https://user-images.githubusercontent.com/52176616/122620872-0ee20d00-d062-11eb-828b-b4c16dc03b0d.png)

kgrewal26 replied:
Sounds good - I do remember having trouble with the .ioc file in the past but I'll let you know asap if I run into troubles

Pop0097 reviewed (commented):
Thanks for getting this in. I have a few comments. 

Do you have a dev board that you can test this on?

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -7,9 +7,9 @@
-    MX_TIM4_Init();
-    HAL_TIM_Base_Init(&htim4);
-    HAL_TIM_Base_Start(&htim4);
+    MX_TIM12_Init();
```
You don't need this. If you use CubeMX to configure the ioc this function will be called in `./boardfiles/src/main.c`

kgrewal26 commented:
> Thanks for getting this in. I have a few comments.
> 
> Do you have a dev board that you can test this on?

Unfortunately not, is there a way someone could test it for me?

Pop0097 commented:
Here are the people who currently have dev boards. You can ask around and see if someone is willing:

- Aadi
- Dhruv
- Gordon
- Anni
- Anthony 

Your best bet is probably Aadi if I recall correctly

kgrewal26 commented:
> Here are the people who currently have dev boards. You can ask around and see if someone is willing:
> 
> * Aadi
> * Dhruv
> * Gordon
> * Anni
> * Anthony
> 
> Your best bet is probably Aadi if I recall correctly

 I'm not exactly sure how I'd test something like this.... is it safe to assume they'd know?
 
 Also, the module isn't exactly complete until sensor fusion passes me in the GPS struct, and IIRC that hasn't been done yet.

kgrewal26 commented:
A few new things...

Switched over to timer14 (this hasn't been configured in ioc yet!)

Fixed the way I set the recentUTCTime (accounted for HHMMSS.00 format)

And other small little details.

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -0,0 +1,64 @@
+    //check if its new and set the recentUTCTime to it
+    if (gpsData->timeIsNew){
+
+        __HAL_TIM_SET_COUNTER(&htim12, 0);
```
htim14

If you are changing a variable name, a good practice is to always check with CTRL-F if you have changed all instances ;))

Pop0097 commented on Autopilot/Src/timeStampingPOGI.cpp:
```diff
@@ -1,24 +1,20 @@
-    MX_TIM12_Init();
-    HAL_TIM_Base_Init(&htim12);
-    HAL_TIM_Base_Start(&htim12);
+    HAL_TIM_Base_Init(&htim14);
+    HAL_TIM_Base_Start(&htim14);
```
tests failing because htim14 not present. Will be fixed once we set up ioc file :))
